apps/src/netsim/NetSimRouterLogModal.js

```diff
@@ -148,28 +171,190 @@ NetSimRouterLogModal.prototype.render = function () {
-      }
-  }.bind(this));
-
   this.rootDiv_.find('th').click(function (event) {
     this.onSortHeaderClick_($(event.target).attr('data-sort-key'));
   }.bind(this));
+
+  // Add rows to the table
+  var rows = this.getSortedFilteredLogEntries(this.logEntries_)
+      .slice(0, MAXIMUM_ROWS_IN_FULL_RENDER)
+      .map(this.makeTableRow_.bind(this));
+  this.rootDiv_.find('tbody').append(rows);
+
+  if (rows.length === MAXIMUM_ROWS_IN_FULL_RENDER) {
+    var maxRenderedWarning = document.createElement('div');
+    maxRenderedWarning.className = 'log-browser-limit-message';
+    maxRenderedWarning.innerText = i18n.showingFirstXLogEntries({
+      x: MAXIMUM_ROWS_IN_FULL_RENDER
+    });
+    this.rootDiv_.find('table').after(maxRenderedWarning);
+  }
+};
+
+/**
+ * Convert the given set of log entries to table rows and insert them
+ * into the DOM, instead of re-rendering the whole table.
+ * @param {!NetSimLogEntry[]} newEntries
+ * @private
+ */
+NetSimRouterLogModal.prototype.renderNewLogEntries_ = function (newEntries) {
+  // Be lazy, don't render at all if not visible.
+  if (!this.isVisible() || newEntries.length === 0) {
+    return;
+  }
+
+  /** @type {jQuery} Table body element. */
+  var tbody = this.rootDiv_.find('tbody');
+
+  // Get existing table row elements, which are already sorted and filtered.
+  var oldRows = tbody.find('tr');
+
+  // Sort and filter the new entries, and generate DOM rows for them.
+  newEntries = this.getSortedFilteredLogEntries(newEntries);
+  var newRows = $(newEntries.map(this.makeTableRow_.bind(this)));
+
+  // Get the current sort value function
+  var iterateeFunction = NetSimRouterLogModal.iterateeMap[this.sortBy_];
+
+  // Walk both collections to merge new rows into the DOM
+  var nextOld = getNextInfo(oldRows, 0, iterateeFunction);
+  var nextNew = getNextInfo(newRows, 0, iterateeFunction);
+  var insertHere = false;
+  while (nextNew.index < newRows.length && nextOld.index < oldRows.length) {
+
+    // Is this where the next row goes?
+    insertHere = nextNew.sortValue < nextOld.sortValue ?
+        !this.sortDescending_ : this.sortDescending_;
+
+    if (insertHere) {
+      nextNew.tableRow.insertBefore(nextOld.tableRow);
+      nextNew = getNextInfo(newRows, nextNew.index + 1, iterateeFunction);
+    } else {
+      nextOld = getNextInfo(oldRows, nextOld.index + 1, iterateeFunction);
+    }
+  }
+
+  // Put whatever's left on the end of the table
+  tbody.append(newRows.slice(nextNew.index));
+};
+
+/**
+ * Generates a helper object for performing the log row merge.
+ * @param {jQuery} rows - Wrapped collection of table rows.
+ * @param {!number} atIndex - Index into `rows` at which info should be generated.
+ * @param {!function(NetSimLogEntry)} iterateeFunction - function to get current
+ *        sort value from log entry object.
+ * @returns {{index: number, tableRow: jQuery, sortValue: ?}}
+ */
+function getNextInfo(rows, atIndex, iterateeFunction) {
+  var tempRow = rows.eq(atIndex);
```

Comment:
Would be nice if there is a way to test this!   Given the dependency on the DOM, is there any way to do that that wouldn't be too difficult?   I'm interested in cases like:
- added rows at end of the table
- no rows added
- initially empty table
- etc.